src/ReservoirRouter.sol

```diff
@@ -0,0 +1,291 @@
+
+import { PeripheryImmutableState } from "src/abstract/PeripheryImmutableState.sol";
+import { PeripheryPayments } from "src/abstract/PeripheryPayments.sol";
+import { PredicateHelper } from "src/abstract/PredicateHelper.sol";
+import { Multicall } from "src/abstract/Multicall.sol";
+
+contract ReservoirRouter is
+    IReservoirRouter,
+    PeripheryImmutableState,
+    PeripheryPayments,
+    PredicateHelper,
+    Multicall
+{
+    uint256 constant MINIMUM_LIQUIDITY = 1e3;
+
+    constructor (address aFactory, address aWETH) PeripheryImmutableState(aFactory, aWETH)
+    {}
+
+    function _addLiquidity(
+        address tokenA,
+        address tokenB,
+        uint curveId,
+        uint amountADesired,
+        uint amountBDesired,
+        uint amountAMin,
+        uint amountBMin
+    ) private returns (uint amountA, uint amountB, address pair) {
+        pair = factory.getPair(tokenA, tokenB, curveId);
+
+        if (pair == address(0)) {
+            pair = factory.createPair(tokenA, tokenB, curveId);
+        }
+
+        (uint256 reserveA, uint256 reserveB) = ReservoirLibrary.getReserves(address(factory), tokenA, tokenB, curveId);
+        if (reserveA == 0 && reserveB == 0) {
+            (amountA, amountB) = (amountADesired, amountBDesired);
+        }
+        else {
+            uint amountBOptimal = ReservoirLibrary.quote(amountADesired, reserveA, reserveB);
+            if (amountBOptimal <= amountBDesired) {
+                require(amountBOptimal >= amountBMin, "RR: INSUFFICIENT_B_AMOUNT");
+                (amountA, amountB) = (amountADesired, amountBOptimal);
+            } else {
+                uint amountAOptimal = ReservoirLibrary.quote(amountBDesired, reserveB, reserveA);
+                assert(amountAOptimal <= amountADesired);
+                require(amountAOptimal >= amountAMin, "RR: INSUFFICIENT_A_AMOUNT");
+                (amountA, amountB) = (amountAOptimal, amountBDesired);
+            }
+        }
+    }
+
+    function addLiquidity(
+        address tokenA,
+        address tokenB,
+        uint256 curveId,
+        uint amountADesired,
+        uint amountBDesired,
+        uint amountAMin,
+        uint amountBMin,
+        address to
+    ) external payable returns (uint256 amountA, uint256 amountB, uint256 liquidity) {
+        address pair;
+        (amountA, amountB, pair) = _addLiquidity(tokenA, tokenB, curveId, amountADesired, amountBDesired, amountAMin, amountBMin);
+
+        _pay(tokenA, msg.sender, pair, amountA);
+        _pay(tokenB, msg.sender, pair, amountB);
+
+        liquidity = IReservoirPair(pair).mint(to);
+    }
+
+    function removeLiquidity(
+        address tokenA,
+        address tokenB,
+        uint256 curveId,
+        uint256 liquidity,
+        uint256 amountAMin,
+        uint256 amountBMin,
+        address to // when withdrawing to the router in UniV3, this address is 0. But we can just put the router's address
```

Comment:
This comment doesn't make sense to me, if we set the address to zero we will destroy the users liquidity?

Reply:
oh yeah this is just a note to myself about UniV3's impl. Regarding this problem, I was thinking, we can add a `require(to != address(0)` as a safety check for all swap and `removeLiquidity`.

What do you think?

Reply:
Happy to add the safety check, we can always decide to remove if we really want to squeeze gas efficiency. (keep in mind some ERC20s check for transfer to zero address anyway)

Reply:
okay added for now.

Did some research. USDC has this check to make sure we don't transfer to the 0 address. 

USDT doesn't (and the balance of the 0 address is non-zero), which means this has happened before lol